ntp-proto/src/filter.rs

```diff
@@ -299,6 +328,124 @@ impl Peer {
 
         true
     }
+
+    #[allow(dead_code)]
+    fn update_with_packet(
+        &mut self,
+        local_clock_time: NtpTimestamp,
+        system_precision: NtpDuration,
+        mut packet: NtpHeader,
+        destination_timestamp: NtpTimestamp,
+    ) -> Option<FilterTuple> {
+        // we map stratum 0 (unspecified) to MAXSTRAT to make stratum
+        // comparisons simpler and to provide a natural interface
+        // for radio clock drivers that operate for convenience at stratum 0.
+        if packet.stratum == 0 {
+            packet.stratum = MAX_STRATUM;
+        }
+
+        self.last_packet = packet;
+
+        // Verify the server is synchronized with valid stratum and
+        // reference time not later than the transmit time.
+        if !self.last_packet.leap.is_synchronized() || self.last_packet.stratum >= MAX_STRATUM {
+            // this peer is unsynchronized
+            return None;
+        }
+
+        // verify root distance
+        let packet_dispersion =
+            self.last_packet.root_delay / 2i64 + self.last_packet.root_dispersion;
+        let time_travel =
+            self.last_packet.reference_timestamp > self.last_packet.transmit_timestamp;
+        if packet_dispersion >= NtpDuration::MAX_DISPERSION || time_travel {
+            return None; /* invalid header values */
+        }
+
+        // host_poll
+        let poll_interval = self.host_poll;
+        self.poll_update(local_clock_time, poll_interval);
+        self.reach.update();
+
+        let r = &self.last_packet;
+        let packet_precision = NtpDuration::from_exponent(r.precision);
+
+        let tuple = if let crate::packet::NtpAssociationMode::Broadcast = r.mode {
+            let offset = r.transmit_timestamp - destination_timestamp;
+            let delay = BROADCAST_DELAY;
+            let dispersion =
+                packet_precision + system_precision + multiply_by_phi(BROADCAST_DELAY * 2i64);
+
+            FilterTuple {
+                offset,
+                delay,
+                dispersion,
+                time: local_clock_time,
+            }
```

Comment:
Lets not deal with broadcast yet.

Reply:
finding all the places where we do need to add it later will be a nightmare. When the cost is small, I figure we can just do the implementation. Or leave an explicit `todo!()`